### ISSUE TITLE
Wrong width for certain characters

### ISSUE THREAD

**Issue by nu8** [vim/vim#6778]
Using this:

<https://github.com/vim/vim-win32-installer/releases/download/v8.2.1505/gvim_8.2.1505_x64.zip>

All of these characters (and perhaps others):

ℹ ⌨ ⏏ ⏭ ⏮ ⏯ ⏱ ⏲ ⏸ ⏹
⏺ ☀ ☁ ☂ ☃ ☄ ☎ ☑ ☝ ☠
☠️ ☢ ☣ ☦ ☪ ☮ ☯ ☸ ☹ ♟
♨ ♻ ♾ ⚒ ⚔ ⚕ ⚕️ ⚖ ⚖️ ⚗
⚙ ⚛ ⚜ ⚠ ⚰ ⚱ ⛈ ⛏ ⛩ ⛰
⛱ ⛴ ⛷ ⛸ ✂ ✈ ✈️ ✉ ✍ ✏
✒ ✔ ✖ ✝ ✡ ✳ ✴ ❄ ❇ ❣
❤ ❤️ ➡ 🅰 🅱 🅾 🅿

are "width 1" instead of "width 2". This causes overlapping and other problems.
I notice if I open the same file in Windows Notepad, output is correct.

**Comment by brammool** [comment]
I am confused by the code.  We have two tables: emoji_all, which is used to classify characters as emoji and label them as "ambiguous width".  But when getting the cell size another table is used, for which I cannot find a reason why these emoji characters are different. When I use the emoji_all table to decide whether a character is double width, it appears to work better, especially in the GUI.  So let me do that know, and await complaints what's broken now...

**Comment by brammool** [comment]
That doesn't work well, some characters marked as emoji do not really show double-width.
I guess the table https://unicode.org/Public/emoji/1.0/emoji-data.txt
lists the double-wide characters with "emoji" and the single width ones as "text".

**Comment by nu8** [comment]
probably not helpful, but I know that this program works as expected with the same inputs from the original post:

https://github.com/rizonesoft/Notepad3

**Comment by puremourning** [comment]
I'm not sure if this is super helpful or not, but I found this paper describing a character property for actual double width characters: http://www.unicode.org/reports/tr11/tr11-36.html

> The Unicode character property East_Asian_Width provides a default classification of characters, which an implementation can use to decide at runtime whether to treat a character as narrow or wide

> The East_Asian_Width is an informative property and provides a useful concept for implementations that

> Have to interwork with East Asian legacy character encodings
Support both East Asian and Western typography and line layout
Need to associate fonts with unmarked text runs containing East Asian characters

Not sure if that's related to the emoji characters too. Sorry I didn't read it in detail, I just happened to be researching something related today.

**Comment by jamessan** [comment]
> I am confused by the code. We have two tables: emoji_all, which is used to classify characters as emoji and label them as "ambiguous width". But when getting the cell size another table is used, for which I cannot find a reason why these emoji characters are different.

There was a Unicode release (9, I think) that changed the width of emoji.  At the time, I think you decided to use the old width for the existing emoji and the new width for newly introduced emoji

**Comment by chrisbra** [comment]
do we need to update the unicode tables? I believe there was a new release around march and another update/release coming at the end of this year(?)

**Comment by brammool** [comment]
Christian wrote:

> do we need to update the unicode tables? I believe there was a new
> release around march and another update/release coming at the end of
> this year(?)

You can run the script and check if differences make sense.
I haven't tried it yet.

-- 
press CTRL-ALT-DEL for more information

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
Looks like the file just doesn't have the information.  For example 0x23f1 (stopwatch) is clearly double width in the font.
But in https://unicode.org/Public/emoji/1.0/emoji-data.txt it is marked as "text", thus would be single width.
In the later https://unicode.org/Public/emoji/12.1/emoji-data.txt it is given Extended_Pictographic, but so are emoji's that are single width.  The very similar 0x23f0 (alarm clock) gets Emoji_Presentation.
Thus it looks like these tables are not useful to decide whether an emoji is single or double width.

**Comment by brammool** [comment]
I have improved this a bit, mainly based on just trying which characters appear to be double width.
I have added the setcellwidths() function so you can tune this to your system.  I could not find a Unicode document that specifies the widths, thus it depends on your environment.
I also tried to make the 0xfe0f modifier work, but it appears I don't have a way to display the characters properly (with GTK), which makes this impossible for me to make work.  Perhaps someone can work on this, but it's not going to be easy, since many places in Vim assume the width only depends on the base character, not combining characters that follow.

**Comment by nu8** [comment]
https://github.com/equalsraf/neovim-qt/issues/735

**Comment by chrisbra** [comment]
so let's close this then.

**Comment by k-takata** [comment]
Refs:
https://github.com/vim/vim/commit/08aac3c6192f0103cb87e280270a32b50e653be1
https://github.com/vim/vim/commit/4e4473c927167fd24e5c8df90e0e8035080cf2da

**Comment by nu8** [comment]
So, I am kind of confused why this was closed. Has the issue been fixed? If not, is it being closed as "WONTFIX"?

Cause I just tried again with the characters in my original post, and they still seem broken.

Do I need to mess with `setcellwidths`? If so, what would that even look like? I would appreciate some clarification.

**Comment by ychin** [comment]
@nu8 I think Bram fixed some of the issue but it's not fully resolved, and `setcellwidths` is a way for you to manually set each unicode chars to be width 2 to get around it for now, but it's not ideal.

First, some emoji's have both textual and "emoji" representation, but default to textual. E.g. ⏏ or ™. You can see the full list at https://unicode.org/emoji/charts/text-style.html (it's the rightmost column. See http://xahlee.info/comp/text_vs_emoji.html for explanation). For these, I think Vim now still correct treats those as width 1. That said, it may be system-dependent as some systems may always draw these emojis as "emoji" representation and ignore the default. It's difficult for Vim to know that.

You can also use 0xfe0e (text variant) or 0xfe0f (emoji variant) to mandate the emoji's representation (your example includes some of those). E.g. ⏏︎ is the textual rep, while ⏏️ is the emoji rep (you may have to use the right browser to see it. You can always copy the symbols to Vim and then do `ga` on the character to see the actual code). Vim doesn't currently correctly set the width for these characters per https://github.com/vim/vim/issues/6778#issuecomment-683289089. I feel like it's worth it to fix but looking at the code it does look like it would involve a bit of refactoring.

Just some potential future solutions I can think of:

1. Currently the `'emoji'` settings is quite misleading, as it says "all Unicode emoji characters are considered to be full width", which is not true. It only treats the ones under `emoji_wide` to be full width. Maybe we should fix it to be a little more clear what it does.
2. Make a new setting (e.g. `'emojirep'`) that allows forcing all emojis to use the "emoji" representation and use double width. This way all emojis will use the pictures (we may have to inject `fe0f` composing chars to the text rendering) version, and will be double-width. Current thinking is the setting could take on 3 values: `default`, `text`, `emoji`.
3. Support the "emoji" variant (`FE0F`) composing character and dynamically switch to width 2, i.e. ⏏︎ vs ⏏️. This is more work but without it Vim will continue to draw emojis incorrectly unless we brute force it and use `emojirep=emoji` as suggested above (which isn't always a good thing because ™ is usually in textual form and we don't want to override it to ™️).
4. Implement system-dependent behavior instead of forcing users to use `setcellwidth`? Maybe on known systems (Windows?) that ignore textual representation on certain emojis, select those as double width. This seems annoying though.

I could work on some of those fixes, but wanted to see what @brammool and others think first.

**Comment by ychin** [comment]
Also, I'm assuming "text" variants are single-width, and "emoji" variants are double-width. I still don't know how to confirm that or whether that is even meaningful definition within Unicode.

**Comment by nu8** [comment]
> 2. Make a new setting (e.g. `'emojirep'`) that allows forcing all emojis to
> use the "emoji" representation and use double width. This way all emojis will
> use the pictures (we may have to inject `fe0f` composing chars to the text
> rendering) version, and will be double-width. Current thinking is the setting
> could take on 3 values: `default`, `text`, `emoji`.

I think this would be the correct solution, or at least the start of a correct
one. It is possible that Windows in general, or maybe just some Windows programs
ignore the default as you say. So if and when that happens, Vim thinks a
character is being represented in the default, while the program is ignoring the
default and doing the emoji representation.

If the default is emoji, then it doesnt matter. But if the default is text then
its a problem. So a good compromise for users who wish to opt in, would be just
force emoji display and double width for all suspect characters.

**Comment by tonymec** [comment]
see
```vim
    :help 'ambiwidth'
    :help setcellwidths()
" and also
    :new
    :source $VIMRUNTIME/tools/emoji_list.vim
```

Best regards,
Tony.

**Comment by brammool** [comment]
> @nu8 I think Bram fixed some of the issue but it's not fully resolved,
> and `setcellwidths` is a way for you to manually set each unicode
> chars to be width 2 to get around it for now, but it's not ideal.
> 
> First, some emoji's have both textual and "emoji" representation, but
> default to textual. E.g. ⏏ or ™. You can see the full list at
> https://unicode.org/emoji/charts/text-style.html (it's the rightmost
> column. See http://xahlee.info/comp/text_vs_emoji.html for
> explanation). For these, I think Vim now still correct treats those as
> width 1. That said, it may be system-dependent as some systems may
> always draw these emojis as "emoji" representation and ignore the
> default. It's difficult for Vim to know that.
>
> You can also use 0xfe0e (text variant) or 0xfe0f (emoji variant) to
> mandate the emoji's representation (your example includes some of
> those). E.g. ⏏︎ is the textual rep, while ⏏️ is the emoji rep (you may
> have to use the right browser to see it. You can always copy the
> symbols to Vim and then do `ga` on the character to see the actual
> code). Vim doesn't currently correctly set the width for these
> characters per
> https://github.com/vim/vim/issues/6778#issuecomment-683289089. I feel
> like it's worth it to fix but looking at the code it does look like it
> would involve a bit of refactoring.

Vim does currently not support the representation modifier.  I had a
look at it, but it's going to be a lot of work to implement, since Vim
assumes the width is given by the base character, but now it depends on
what follows.  This will require changes all over the place.

> Just some potential future solutions I can think of:
> 
> 1. Currently the `'emoji'` settings is quite misleading, as it says
> "all Unicode emoji characters are considered to be full width", which
> is not true. It only treats the ones under `emoji_wide` to be full
> width. Maybe we should fix it to be a little more clear what it does.

I'll update the explanation.

> 2. Make a new setting (e.g. `'emojirep'`) that allows forcing all
> emojis to use the "emoji" representation and use double width. This
> way all emojis will use the pictures (we may have to inject `fe0f`
> composing chars to the text rendering) version, and will be
> double-width. Current thinking is the setting could take on 3 values:
> `default`, `text`, `emoji`.

I don't think this will be useful, since there will still be characters
that are displayed wrong.  It may look better, but if some character is
still wrong the setting won't help.  You could only switch between
broken-one-way and broken-another-way.

> 3. Support the "emoji" variant (`FE0F`) composing character and
> dynamically switch to width 2, i.e. ⏏︎ vs ⏏️. This is more work but
> without it Vim will continue to draw emojis incorrectly unless we
> brute force it and use `emojirep=emoji` as suggested above (which
> isn't always a good thing because ™ is usually in textual form and we
> don't want to override it to ™️).

You can't brute force it, since some characters will have FE0F and some
don't.  That's the whole idea of the modifier, that you can use both.
As mentioned, adding support for the modifier is going to be a lot of
work.  Also, it will only work if the terminal also supports it.

> 4. Implement system-dependent behavior instead of forcing users to use
> `setcellwidth`? Maybe on known systems (Windows?) that ignore textual
> representation on certain emojis, select those as double width. This
> seems annoying though.

There is no system-dependent behavior but terminal-dependent behavior.
And it will even depend on the font used.

> I could work on some of those fixes, but wanted to see what @brammool
> and others think first.
> 
> Also, I'm assuming "text" variants are single-width, and "emoji"
> variants are double-width. I still don't know how to confirm that or
> whether that is even meaningful definition within Unicode.

I could not find a specification for what characters use what width.  It
appears these characters are really meant to be used in a proportional
font.

-- 
The Law of VIM:
For each member b of the possible behaviour space B of program P, there exists
a finite time t before which at least one user u in the total user space U of
program P will request b becomes a member of the allowed behaviour space B'
(B' <= B).
In other words: Sooner or later everyone wants everything as an option.
                                        -- Vince Negri

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by ychin** [comment]
> > 2. Make a new setting (e.g. `'emojirep'`) that allows forcing all
 emojis to use the "emoji" representation and use double width. This
 way all emojis will use the pictures (we may have to inject `fe0f`
 composing chars to the text rendering) version, and will be
 double-width. Current thinking is the setting could take on 3 values:
 `default`, `text`, `emoji`.
>
> I don't think this will be useful, since there will still be characters
that are displayed wrong.  It may look better, but if some character is
still wrong the setting won't help.  You could only switch between
broken-one-way and broken-another-way.

---

I think `set emojirep=emoji` would mostly be harmless. It makes all emojis double-width. On systems that don't support the emoji/text variants, it would just make them take up more space. On system that do support the variant, it will make the emojis render in their emoji variant and be double-width. It's basically Vim throwing its hands up and taking the most conservative approach to avoid truncation.

I did test this on macOS's Terminal by hacking `$VIMRUNTIME/tools/emoji_list.vim` by appending `FE0F` to all of the emojis, and forcing `utf_char2cells` to return 2 on any emojis. That actually worked perfectly and display emoji variants / and have full width:

**Before:**

![image](https://user-images.githubusercontent.com/1217449/92873679-277b1380-f3bc-11ea-82c6-9d036e210fb4.png)

(Text for reference)
> |‼| 1  |⏭| 2 |⚙| 1
> |⁉| 1  |⏮| 2 |⚛| 1
> |™| 1  |⏯| 2 |⚜| 1
> |ℹ| 1  |⏰| 2 |⚠| 1
> |↔| 1  |⏱| 2 |⚡| 2
> |↕| 1  |⏲| 2 |⚪| 2
> |↖| 1  |⏳| 2 |⚫| 2
> |↗| 1  |⏸| 2 |⚰| 1
> |↘| 1  |⏹| 2 |⚱| 1
> |↙| 1  |⏺| 2 |⚽| 2
> |↩| 1  |Ⓜ| 2 |⚾| 2
> |↪| 1  |▪| 1  |⛄| 2
> |⌚| 2 |▫| 1  |⛅| 2
> |⌛| 2 |▶| 1  |⛈| 2
> |⌨| 1  |◀| 1  |⛎| 2
> |⏏| 1  |◻| 1  |⛏| 2
> |⏩| 2 |◼| 1  |⛑| 2
> |⏪| 2 |◽| 2 |⛓| 2
> |⏫| 2 |◾| 2 |⛔| 2
> |⏬| 2 |☀| 1  |⛩| 2


**After (forcing everything to emoji rep):**

![image](https://user-images.githubusercontent.com/1217449/92873217-a9b70800-f3bb-11ea-8591-71a4970cbc9a.png)

(Text for reference)
> |‼️| 2 |⏭️| 2 |⚙️| 2
> |⁉️| 2 |⏮️| 2 |⚛️| 2
> |™️| 2 |⏯️| 2 |⚜️| 2
> |ℹ️| 2 |⏰️| 2 |⚠️| 2
> |↔️| 2 |⏱️| 2 |⚡️| 2
> |↕️| 2 |⏲️| 2 |⚪️| 2
> |↖️| 2 |⏳️| 2 |⚫️| 2
> |↗️| 2 |⏸️| 2 |⚰️| 2
> |↘️| 2 |⏹️| 2 |⚱️| 2
> |↙️| 2 |⏺️| 2 |⚽️| 2
> |↩️| 2 |Ⓜ️| 2 |⚾️| 2
> |↪️| 2 |▪️| 2 |⛄️| 2
> |⌚️| 2 |▫️| 2 |⛅️| 2
> |⌛️| 2 |▶️| 2 |⛈️| 2
> |⌨️| 2 |◀️| 2 |⛎️| 2
> |⏏️| 2 |◻️| 2 |⛏️| 2
> |⏩️| 2 |◼️| 2 |⛑️| 2
> |⏪️| 2 |◽️| 2 |⛓️| 2
> |⏫️| 2 |◾️| 2 |⛔️| 2
> |⏬️| 2 |☀️| 2 |⛩️| 2

Tested on Windows (on Windows Terminal) though, and it doesn't seem to respect `FE0F` at all 😔. Seems like it will just pick one variant and stick with it, ignoring the variant marker. So this option will just guarantee the emojis take double-width.

But yeah it's not perfect. We don't have control how these texts will be rendered, and proper emoji (really it's Unicode) support on different OSes are still inconsistent. Also, some emojis like the trademark symbol (™️) are usually supposed to show up in textual variant instead of emoji variant, which this option will force it to be double-width. I still think there is some value in this though. It gives some option for you to control the spacing of the emojis which could make it less annoying if you happen to be editing a file with lots of them and seeing inconsistent widths and truncation.

Maybe the name I suggested is also non-ideal. This is almost like `ambiwidth` on steroids, for emojis only (ambiwidth is more for ambiguous characters in certain languages).

---

`set emojirep=text` would probably be a little more work. Not all emojis have textual representation (but they *are* defined in the spec), so maybe the setting is not as useful as the other option.
